bridge-ffi/src/lib.rs

```diff
@@ -0,0 +1,650 @@
+struct UserDataHandle(*mut libc::c_void);
+unsafe impl Send for UserDataHandle {}
+unsafe impl Sync for UserDataHandle {}
+
+lazy_static::lazy_static! {
+    static ref DEFAULT_INIT_RESPONSE_BYTES: tmprl_bytes_t = {
+        tmprl_bytes_t::from_vec_disable_free(bridge::InitResponse::default().encode_to_vec())
+    };
+
+    static ref DEFAULT_SHUTDOWN_RESPONSE_BYTES: tmprl_bytes_t = {
+        tmprl_bytes_t::from_vec_disable_free(bridge::ShutdownResponse::default().encode_to_vec())
+    };
+
+    static ref DEFAULT_REGISTER_WORKER_RESPONSE_BYTES: tmprl_bytes_t = {
+        tmprl_bytes_t::from_vec_disable_free(bridge::RegisterWorkerResponse::default().encode_to_vec())
+    };
+}
+
+/// A runtime owned by Core. This must be passed to tmprl_runtime_free when no
+/// longer in use. This must not be freed until every call to every tmprl_core_t
+/// instance created with this runtime has been shutdown.
+pub struct tmprl_runtime_t {
+    tokio_runtime: std::sync::Arc<tokio::runtime::Runtime>,
+}
+
+/// Create a new runtime. The result is never null and must be freed via
+/// tmprl_runtime_free when no longer in use.
+#[no_mangle]
+pub extern "C" fn tmprl_runtime_new() -> *mut tmprl_runtime_t {
+    Box::into_raw(Box::new(tmprl_runtime_t {
+        // TODO(cretz): Options to configure thread pool?
+        tokio_runtime: std::sync::Arc::new(
+            tokio::runtime::Builder::new_multi_thread()
+                .enable_all()
+                .build()
+                .unwrap(),
+        ),
+    }))
+}
+
+/// Free a previously created runtime.
+#[no_mangle]
+pub extern "C" fn tmprl_runtime_free(runtime: *mut tmprl_runtime_t) {
+    if !runtime.is_null() {
+        unsafe {
+            Box::from_raw(runtime);
+        }
+    }
+}
+
+/// A core instance owned by Core. This must be passed to tmprl_core_shutdown
+/// when no longer in use which will free the resources.
+pub struct tmprl_core_t {
+    tokio_runtime: std::sync::Arc<tokio::runtime::Runtime>,
+    // TODO(cretz): Any concerns with the dynamic dispatch overhead here?
+    // TODO(cretz): Should we ask Rust SDK to give us back a struct instead?
+    // TODO(cretz): We could use generics, but impl traits don't cross the FFI
+    // boundary properly
+    core: std::sync::Arc<dyn temporal_sdk_core_api::Core>,
+}
+
+/// Callback called by tmprl_core_init on completion. The first parameter of the
+/// callback is user data passed into the original function. The second
+/// parameter is a core instance if the call is successful or null if not. If
+/// present, the core instance must be freed via tmprl_core_shutdown when no
+/// longer in use. The third parameter of the callback is a byte array for a
+/// InitResponse protobuf message which must be freed via tmprl_bytes_free.
+type tmprl_core_init_callback = unsafe extern "C" fn(
+    user_data: *mut libc::c_void,
+    core: *mut tmprl_core_t,
+    resp: *const tmprl_bytes_t,
+);
+
+// TODO(cretz): Catch unwind for badly behaved callbacks?
```

Comment:
I don't think we need to -- lang should never panic in its callback handlers. If we prevent it we are likely in a broken state anyway

Reply:
Removed this TODO. I shouldn't have put it here btw, it is the Rust code where I was wondering if we should catch unwind, but per discussion we want it to crash.